src/FakeItEasy/ParameterInfoExtensions.cs

```diff
@@ -0,0 +1,27 @@
+namespace FakeItEasy
+{
+    using System.Linq;
+    using System.Reflection;
+
+    /// <summary>
+    /// Provides extension methods for <see cref="ParameterInfo"/>.
+    /// </summary>
+    internal static class ParameterInfoExtensions
+    {
+        public static bool IsOutOrRef(this ParameterInfo parameterInfo)
+        {
+            if (!parameterInfo.ParameterType.IsByRef)
+            {
+                return false;
+            }
+
+#if FEATURE_READONLY_IN_PARAMETERS
+            var parameterAttributes = parameterInfo.CustomAttributes;
+            return parameterAttributes == null ||
+                   parameterAttributes.All(customAttributeData => customAttributeData.AttributeType.FullName != "System.Runtime.CompilerServices.IsReadOnlyAttribute");

```

Comment:
If I understand correctly, the `#if` is just an optimization; platforms that don't have `IsReadOnlyAttribute` will just never see the attribute. Right?
I wonder if it's worth keeping the `#if`. Projects that target net471+ should reference the net45 version of FakeItEasy, but if for some reason they reference the net40 version (e.g. because the assembly was referenced manually instead of via nuget, or because of a transitive reference), they won't get the correct behavior.

Reply:
Honestly, I wasn't sure. Even though `ReadOnlyAttribute` is only available in .NET Framework 4.7.1+ and .NET Core 2.0+, it's clearly present on the parameter when we build to target .NET Framework 4.5 and .NET Standard 1.0, and run under .NET Framework 4.6.1 and netcoreapp 1.0. Of course it was put there by a super-modern compiler.
Regardless, the real reason I put the `#if` in was because in net40, the custom attributes look all different. I couldn't compile without it. When I combined that with the uncertainty over the value in trying to support `in` on net40, I figured I'd throw the `#if` in just to see how well it worked on the more modern platforms.